### ISSUE TITLE
propagate type context through `await`

### ISSUE THREAD

**Issue by styfore** [astral-sh/ty#3158]
### Summary

# Summary

## playground link
https://play.ty.dev/dbe21060-c0e8-4cec-96bc-d7d6565d121b

ty 0.0.26 reports a false positive `invalid-assignment` when an async function inside a class return a list[Self]. Works fin if just returning Self

## how to reproduce

```python
from typing import Self

class Parent:
    async def get_list(self) -> list[Self]:
        return [self]

    async def test(self):
        my_list: list[Parent] = await self.get_list()

class Child(Parent):
    async def func2(self):
        childs: list[Child] = await self.get_list()
```

### Version

ty 0.0.26

**Comment by carljm** [comment]
I don't think these are false positives. Pyright and zuban emit the same diagnostics. (Pyrefly and mypy do not.)

If we call `.test()` on an instance of `Child`, we will have a `list[Child]` being assigned to a `list[Parent]` inside that method; this is not sound.

Similarly if we have a `class GrandChild(Child): ...` and we call `.func2()` on an instance of it, we will have a `list[GrandChild]` being assigned to a `list[Child]`.

Of course none of this causes practical unsoundness in this particular example, but that's only because the return value of `get_list()` here is a brand-new ephemeral list, so nothing relies on it being soundly typed (i.e. callers could safely treat it as covariant rather than invariant). But the callers cannot know that.

It does seem like there's a bug in ty here, though, because we stop emitting these diagnostics if the methods become sync instead of async (all other type-checkers have consistent results between the sync and async versions):

```py
from typing import Self

class Parent:
    def get_list(self) -> list[Self]:
        return [self]

    def test(self):
        my_list: list[Parent] = self.get_list()

class Child(Parent):
    def func2(self):
        childs: list[Child] = self.get_list()
```

I suspect (but haven't verified) that that's because bidirectional checking is doing something incorrect here, and our bidirectional checking doesn't see through the awaits?

**Comment by carljm** [comment]
I think my last comment was wrong; thanks @ibraheemdev for questioning it.

It's instructive to de-sugar `Self` to a function-scoped type variable, which is how the spec says it should be treated, and how we treat it:

```py
class Parent:
    def test(self):
        my_list: list[Parent] = get_list(self)

class Child(Parent):
    def func2(self):
        childs: list[Child] = get_list(self)

def get_list[T: Parent](self: T) -> list[T]:
    return [self]
```

All type checkers (including ty) are happy with this version. The `self` argument to `get_list` provides only a lower bound on `T`, so the call-site is free to use type context to choose a wider solution (e.g. `T = Parent`), as long as it satisfies the upper bound on `T`. Similarly in the original version with `Self`, the upper bound on `Self@get_list` is `Parent` (this is determined by the fact that it is defined on `Parent`), and the actual `self` argument (of type `Self@test` or `Self@func2`) provides only a lower bound, so the callers can use type context to widen this.

The mistake I made in my previous analysis was that (if `Self` is a function-scoped type variable) it is not a valid implementation of `get_list` to return a list whose type is constrained in ways the caller can't see. The callers can see all the constraints (the upper bound on `Self@get_list` and the actual call arguments) and are free to use type context to pick any solution that satisfies them.

Taking the above de-sugared repro, it also starts failing if it becomes async:

```py
class Parent:
    async def test(self):
        my_list: list[Parent] = await get_list(self)

class Child(Parent):
    async def func2(self):
        childs: list[Child] = await get_list(self)

async def get_list[T: Parent](self: T) -> list[T]:
    return [self]
```

So bidirectional checking here is working correctly (in the sync version); the bug is that it doesn't yet see through `await`.
